python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +98,434 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
+                compilation of this form. Run ``ffcx --help`` at the
+                command line to see all available options.
+            jit_options: Options used in CFFI JIT compilation of C
+                code generated by FFCx. See ``python/dolfinx/jit.py``
+                for all available options. Takes priority over all other
+                option values.
+            snes_options: Options to pass to the PETSc SNES object.
+        """
+
+        self._u = u
+        self._snes, self._x = create_snes_solver(
+            F, self._u, bcs, J, P, assembly_type, form_compiler_options, jit_options
+        )
+        if assembly_type == fem.petsc.AssemblyType.standard:
+            self._copy_function_to_vec = copy_function_to_vec
+            self._copy_vec_to_function = copy_vec_to_function
+        elif assembly_type == fem.petsc.AssemblyType.block:
+            self._copy_function_to_vec = copy_functions_to_block_vec
+            self._copy_vec_to_function = copy_block_vec_to_functions
+        elif assembly_type == fem.petsc.AssemblyType.nest:
+            self._copy_function_to_vec = copy_functions_to_nest_vec
+            self._copy_vec_to_function = copy_nest_vec_to_functions
+        else:
+            raise ValueError("Unsupported Assembly type")
+
+        # Set PETSc options
+        if snes_options is not None:
+            problem_prefix = f"snes_{id(self)}"
+            self._snes.setOptionsPrefix(problem_prefix)
+
+            opts = PETSc.Options()  # type: ignore
+            opts.prefixPush(problem_prefix)
+            for k, v in snes_options.items():
+                opts[k] = v
+            self._snes.setFromOptions()
+            opts.prefixPop()
+
+            # Delete options from handler post setting
+            for k in snes_options.keys():
+                del opts[k]
+
+    def solve(self) -> tuple[PETSc.Vec, int, int]:  # type: ignore
+        """Solve the problem and update the solution in the problem instance.
+
+        Returns:
+            The solution, convergence reason and number of iterations
+        """
+
+        # Move current iterate into the work array.
+        self.copy_function_to_vec(self._u, self._x)
+
+        # Solve problem
+        self._snes.solve(None, self._x)
+
+        converged_reason = self._snes.getConvergedReason()
+        return self._x, converged_reason, self._snes.getIterationNumber()
+
+    def __del__(self):
+        self._snes.destroy()
+        self._x.destroy()
+
+    @property
+    def snes(self) -> PETSc.SNES:  # type: ignore
+        return self._snes
+
+    @property
+    def copy_function_to_vec(self):
+        return self._copy_function_to_vec
+
+    @property
+    def copy_vec_to_function(self):
+        return self._copy_vec_to_function
+
+
+def F_standard(
+    u: dolfinx.fem.Function,
+    residual: dolfinx.fem.Form,
+    jacobian: dolfinx.fem.Form,
+    bcs: list[dolfinx.fem.DirichletBC],
+    snes: PETSc.SNES,  # type: ignore
```

Comment:
Why is `snes` passed? It's not used.

Reply:
Updated to use `_snes` as the solver has to be passed in to adhere with the SNES Residual and Jacobian function.